aips/aip-56.md

```diff
@@ -0,0 +1,213 @@
+K A::M::R<T>(X)   in A::M::R      
+K A::M::R<T>(X)   in A::M::R<T>
+K A::M::R<T>(X)   in A::M::R<T>(X)
+a                 in s1, s2, .., !t1, !t2, ..
+                    iff (a in s1 or a in s2 or ..) 
+                         and not (a in t1) and not (a in t2) and not ..
+```
+
+
+
+Based on containment, two other operators are of relevance: join (denoted as `*`) and subsumption (denoted as `>=`). In set-oriented terms, join is intersection, and subsumption superset. The definition is as follows:
+
+```
+forall s1, s2:     s1 >= s2       <==> (forall a: a in a2 ==> a in a1)
+forall s1, s2, s:  s == s1 * s2   <==> (forall a: a in s <==> a in s1 and a in s2) 
+```
+
+
+#### Runtime Evaluation
+
+At runtime, a stack of saved access specifiers is maintained as well as an active current access specifier. When a function which has access specifiers is entered, the current active set is saved to the stack. Then its value is _joined_ (the `*` operator above) with that of the current function. If the result of the join does not subsume the specifier of the called function, execution aborts:
+
+```
+access_stack.push(active_accces)
+active_access := active_access * function_access
+let call_allowed := active_access.subsumes(function_access)
+if !(active_access >= function_access) {
+    abort
+}
+```
+Notice that we cannot just always abort when entering a function which potentially has disallowed accesses for two reasons:
+
+- The function many not have any specifiers at all. In this case, the `active_access` stays unmodified at function entry, and the actual accesses are checked as they occur.
+- The implementation of subsumption (`>=`) is allowed to be an over-approximation. This stems from the presence of negation in specifiers, for which subsumption check is hard. Thus, we may not be able to decide `active_access >= function_access`. This case is semantically still sound because, similar with functions with no specifiers at all, accesses are still checked as they occur.
+
+During execution, whenever a resource is accessed (move_to, move_from, exists, borrow_global, borrow_global_mut), this operation is checked against the currently active access specifier, and execution aborts if the access is not allowed.
+
+Note that unless trust boundaries are concerned (transactions and public functions), we rarely expect users to write access specifiers, so the stack is not expected to become very deep in average. 
+
+#### Gas Cost
+
+Access specifiers need to be metered, also to prevent their misuse. This is different from how Aptos currently treats paranoid mode. Because this feature adds additional utility to users, metering is justified.
+
+#### Compatibility
+
+Since access specifiers are encoded in the `FunctionHandle` instances which are copied from imported modules, they get outdated when the dependency is upgraded. The compatibility rule for upgrade is that access specifiers can only be constrained, that is `S_v1 >= S_v2 >= ...`
```

Comment:
Not sure if this is something to bring up in the AIP discussion necessarily, but it might be worth writing some guidance and considerations about compatibility (things to be careful about, maybe suggested compatibility patterns) when it comes to access control specifiers. In particular, negation along with wildcards disallows forever accessing structs that don't even exist right now (which is okay, and perhaps even desired in many cases, but we should say more about compatibility rules when we roll this feature out).

If I originally wrote some code like so:

```
module a {
  fun foo() !reads a::* {}
}
```

And later wanted to upgrade the code like so:
```
module a {
  struct A {}
  struct B {}
  fun foo() reads A, !reads B {}
}
```

It would be incompatible, because `A` is in the updated specifier, but not in the original specifier (even when `A` is a non existent entity in the original specifier).

I realized one could use "wrappers" to split up types into allow lists or ban lists in perpetuity, as generic types give a new "namespace" in the resource access specifier language.

```
module a {
  struct DisAllow<T> { wraps: T }
  fun foo() !reads DisAllow {}
}
```

In future versions, one can add new structs to the module, and as long as they are not wrapped with `DisAllow`, they can be read by `foo()` and its callees.

Reply:
This is interesting, perhaps pull this out into an issue?